src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -22,47 +31,124 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics
         {
-            var tagger = buffer.Properties.GetOrCreateSingletonProperty(
-                _uniqueKey, () => new AggregatingTagger(this, buffer));
-            tagger.OnTaggerCreated();
-            return tagger as IAccurateTagger<T>;
+            if (!this.IsEnabled)
+            {
+                return;
+            }
+
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            var editorSnapshot = spanToTag.SnapshotSpan.Snapshot;
+
+            var cancellationToken = context.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+
+            // See if we've marked any spans as those we want to suppress diagnostics for.
+            // This can happen for buffers used in the preview workspace where some feature
+            // is generating code that it doesn't want errors shown for.
+            var buffer = editorSnapshot.TextBuffer;
+            var suppressedDiagnosticsSpans = default(NormalizedSnapshotSpanCollection);
+            buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
+
+            var eventArgs = _diagnosticService.GetDiagnosticsUpdatedEventArgs(
+                workspace, document.Project.Id, document.Id, context.CancellationToken);
+
+            var sourceText = editorSnapshot.AsText();
+            foreach (var updateArg in eventArgs)
+            {
+                ProduceTags(
+                    context, spanToTag, workspace, document, sourceText, editorSnapshot,
+                    suppressedDiagnosticsSpans, updateArg, cancellationToken);
+            }
         }
 
-        private void RemoveTagger(AggregatingTagger tagger, ITextBuffer buffer)
+        private void ProduceTags(
+            TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag,
+            Workspace workspace, Document document, SourceText sourceText, ITextSnapshot editorSnapshot,
+            NormalizedSnapshotSpanCollection suppressedDiagnosticsSpans, UpdatedEventArgs updateArgs, CancellationToken cancellationToken)
         {
-            buffer.Properties.RemoveProperty(_uniqueKey);
+            try
+            {
+                var id = updateArgs.Id;
+                var diagnostics = _diagnosticService.GetDiagnostics(
+                    workspace, document.Project.Id, document.Id, id, false, cancellationToken);
+
+                var isLiveUpdate = id is ISupportLiveUpdate;
+
+                var requestedSpan = spanToTag.SnapshotSpan;
+                var requestedSnapshot = requestedSpan.Snapshot;
+
+                foreach (var diagnosticData in diagnostics)
+                {
+                    if (this.IncludeDiagnostic(diagnosticData))
+                    {
+                        var actualSpan = diagnosticData
+                            .GetExistingOrCalculatedTextSpan(sourceText)
```

Comment:
Chatted with @heejaechang and he confirmed there is still one problem: when we call GetDiagnostics, those spans may be against an older version of the text snapshot than we are querying here. GetExistingOrCalculatedTextSpan may return a span that is past the end of editorSnapshot, which will result in exceptions being thrown. The try/catch surrounding this might catch it...or might not.

The .TranslateTo of course is a no-op which looks like we're trying to deal with this case. We should really just delete that.

Reply:
Good points!  Let me make sure the code is resilient to all this.